[io.metrics] initial contribution

pravussum opened the pull request:
This is the addon counterpart of the core changes from https://github.com/openhab/openhab-core/pull/2133 in the context of https://github.com/openhab/openhab-core/issues/774.

This IO bundle:

- introduces a new REST endpoint /metrics/prometheus
- binds to the core metrics service and 
- delivers the metrics in a [Prometheus](https://prometheus.io/) compatible format 

[prometheus_output.txt](https://github.com/openhab/openhab-addons/files/5845795/prometheus_output.txt)

To test this you have to run a OH core version that includes the changes from [the PR](https://github.com/openhab/openhab-core/pull/2133).

cc @wborn @kaikreuzer

cweitkamp commented:
@pravussum The build failed. Can you please apply the formatter by running `mvn spotless:apply` in you bundle folder?

fwolter commented on bundles/org.openhab.io.metrics/README.md:
```diff
@@ -0,0 +1,109 @@
+
+## Precondition
+
+The openHAB core metrics must be available - a recent enough openHAB version must be used.
+https://github.com/openhab/openhab-core/pull/2133
```
As this is merged, you could specify OH 3.1.0 instead.

fwolter commented on bundles/org.openhab.io.metrics/README.md:
```diff
@@ -0,0 +1,109 @@
+
+The configuration for the metrics service is available in the openHAB UI under Settings | Other Services | Metrics service.
+Support for pull-based monitoring systems (e. g. Prometheus) is always enabled, since it doesn't imply any significant overhead when not used.
+Support for push-based monitoring systems (e. g. InfluxDB) have to be enabled seperately. 
```
```suggestion
Support for push-based monitoring systems (e. g. InfluxDB) have to be enabled separately. 
```

fwolter commented on bundles/org.openhab.io.metrics/README.md:
```diff
@@ -0,0 +1,109 @@
+Support for push-based monitoring systems (e. g. InfluxDB) have to be enabled seperately. 
+
+The following configuration parameters can be set:
+<table>
+  <thead>
+    <tr><th>Config param</th><th>Description</th><th>Default value</th></tr>
+  </thead>
+  <tbody>
+    <tr><td>influxMetricsEnabled</td><td>Enable the Influx (www.influxdata.com) metrics. Further configuration of the InfluxDB instance necessary.</td><td>false</td></tr>
+  </tbody>
+</table>
```
You could use markdown syntax for the tables.

pravussum replied:
afaik it doesn't support col span which I think will greatly improve clarity here.

fwolter replied:
That's right, markdown doesn't support colspan. Are you using colspan here?

pravussum replied:
:rofl: good point. I think I had in the original version, that's why I introduced the HTML table in the first place, but it changed a lot since then. I'll re-markdown it.

fwolter commented on bundles/org.openhab.io.metrics/README.md:
```diff
@@ -0,0 +1,109 @@
+
+### Prometheus
+
+ Once the IO addon is installed and , the Prometheus endpoint will be available under:
```
```suggestion
Once the IO addon is installed, the Prometheus endpoint will be available under:
```

fwolter commented on bundles/org.openhab.io.metrics/src/main/feature/feature.xml:
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
This file should use spaces for indentation.

fwolter commented on bundles/org.openhab.io.metrics/src/main/feature/feature.xml:
```diff
@@ -0,0 +1,11 @@
+<features name="org.openhab.io.metrics-${project.version}" xmlns="http://karaf.apache.org/xmlns/features/v1.4.0">
+	<repository>mvn:org.openhab.core.features.karaf/org.openhab.core.features.karaf.openhab-core/${ohc.version}/xml/features</repository>
+
+	<feature name="openhab-misc-metrics" description="Metrics service" version="${project.version}">
```
```suggestion
	<feature name="openhab-misc-metrics" description="Metrics Service" version="${project.version}">
```

fwolter commented on bundles/org.openhab.io.metrics/src/main/java/org/openhab/io/metrics/MetricsRestController.java:
```diff
@@ -0,0 +1,81 @@
+@JaxrsApplicationSelect("(" + JaxrsWhiteboardConstants.JAX_RS_NAME + "=" + MetricsService.METRICS_APP_NAME + ")")
+@Path("")
+@JSONRequired
+@RolesAllowed({ Role.USER, Role.ADMIN }) // TODO clarify if regular users should be permitted
```
Can you work on this TODO?

pravussum replied:
I hoped to get some corresponding feedback through the CR ;-) Any opinion?

fwolter replied:
I think there is no sensitive data exposed, so I would go for the user and admin.

fwolter commented on bundles/org.openhab.io.metrics/src/main/resources/OH-INF/config/config.xml:
```diff
@@ -0,0 +1,38 @@
+		https://openhab.org/schemas/config-description-1.0.0.xsd">
+	<config-description uri="io:metrics">
+		<parameter name="influxMetricsEnabled" type="boolean">
+			<label>Influx metrics</label>
```
Words in labels should be capitalized (except prepositions and so on). See https://www.openhab.org/docs/developer/bindings/thing-xml.html#formatting-labels-and-descriptions

fwolter commented on bundles/org.openhab.io.metrics/src/main/resources/OH-INF/config/config.xml:
```diff
@@ -0,0 +1,38 @@
+		</parameter>
+		<parameter name="influxPassword" type="text">
+			<label>InfluxDB password</label>
+			<description>The InfluxDB password (no default).</description>
```
You could add `<context>password</context>`.

fwolter commented on bundles/org.openhab.io.metrics/src/main/resources/OH-INF/config/config.xml:
```diff
@@ -0,0 +1,38 @@
+			<label>InfluxDB password</label>
+			<description>The InfluxDB password (no default).</description>
+		</parameter>
+		<parameter name="influxUpdateIntervalInSeconds" type="integer">
```
```suggestion
		<parameter name="influxUpdateIntervalInSeconds" type="integer" unit="s" min="1">
```

fwolter reviewed (changes requested):
Thanks for your contribution! I reviewed your code and here is my feedback.

There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html`.

pravussum commented:
@fwolter thanks for your review! I addressed all of your comments (unless already commented otherwise). 
Regarding the checkstyle warnings: there is a last one left (about the config.xml) I wouldn't know how to get rid of. But I saw a similar warning is present in other IO bundles.

fwolter reviewed (approved):
LGTM

Your commits are squashed automatically when this is merged.

fwolter commented:
Due to lack of review capacity, the rules for new bindings were weakened. New bindings need only the approval of one reviewer instead of two from now on.

kaikreuzer commented:
Hey @pravussum! I just tried out this add-on, but as soon as I install it (tested on 3.1.0.M4), the Main UI disappears and the REST API vanishes as well... And a /metrics endpoint isn't available either :-/ .

Do you have any idea, what might be wrong?

pravussum commented:
Hi @kaikreuzer, I tried to reproduce using a fresh download from https://openhab.jfrog.io/artifactory/libs-milestone-local/org/openhab/distro/openhab/3.1.0.M4/openhab-3.1.0.M4.zip
I can activate the Metrics service - the `/metrics` endpoint is just the root for several possible supported metric systems, `/prometheus` being the only one offered so far (Influx works push based, so it offers no endpoint). So the full URL would be `http://localhost:8080/metrics/prometheus`, which returned the correct data in my case.
Also the Main UI didn't disappear (what exactly does that mean by the way?). 
But I CAN confirm the REST API had issues, which disappeared after a restart of OH. Also, after the restart the Metrics service didn't appear as installed in the Misc section. I'll have to look into it.

pravussum commented:
Let's continue here #10640...
